### ISSUE TITLE
Promise state loss with onRejected

### ISSUE THREAD

**Issue by rbelouin** [mudge/pacta#4]
As defined in pacta README, onRejected is "Identical to Promise#map but only executed when a promise is rejected rather than resolved."

So I can't get why the following tests aren't equivalent :

```
/* test 1 */
var p = new Promise();
var pp = p.map(x => x + 1).map(x => x * 2).map(console.log);
p.resolve(0);
// prints "2"

/* test 2 */
var p = new Promise();
var pp = p.onRejected(x => x + 1).onRejected(x => x * 2).onRejected(console.log);
p.reject(0);
// does not print anything
```

In fact, onRejected looks identical to an hypothetical Promise#forEach, because :

```
/* test 3 */
var p = new Promise();
var pp = p.onRejected(console.log);
p.reject(0);
// prints "0"
```

But :

```
var p = new Promise(42);
var pp = p.onRejected(x => x);
p.value == pp.value; // false because pp.value is not defined

var p = new Promise(); p.reject(42);
var pp = p.onRejected(x => x);
p.reason == pp.reason; // false because pp.reason is not defined
```

I expected onRejected to be chainable, but it does not seems to be so.

**Comment by mudge** [comment]
Hi @rbelouin,

I've been working on Pacta again to provide a complete [ECMAScript 2015](http://www.ecma-international.org/ecma-262/6.0/#sec-promise-objects)-compatible API for Promises (you can see the progress in a [separate branch](https://github.com/mudge/pacta/compare/ecmascript-2015)) and I've been rethinking `mapError` and `chainError`.

I want to deprecate `mapError` as it can be implemented purely in terms of `onRejected`:

```javascript
Promise.prototype.mapError = function (f) {
    return this.onRejected(function (reason) {
        throw f(reason);
    });
};
```

This would reduce the surface area of the foundational algebraic (viz. [Fantasy Land](https://github.com/puffnfresh/fantasy-land)) API to just `map` and `onRejected` (which have some symmetry).

That said, `chainError` *is* needed as the missing API to support functions that return promises being bound to rejection. I might alias this to `chainRejected` to make the relationship between `map`, `onRejected` and `chain`, `chainRejected`.

I'd actually leave `mapError` in the library (perhaps indefinitely) for compatibility reasons but might drop it from the documentation.

Do you have any opinions or objections?

**Comment by rbelouin** [comment]
Hi!

Thank you for asking, I don't have any objection. Dropping it from the docs is a good idea, you can also add some warnings via ```console.warn()``` if it is available.

**Comment by mudge** [comment]
Thanks for responding so quickly, I'm looking into the possibility of using [`util.deprecate`](https://nodejs.org/api/util.html#util_util_deprecate_function_string) as well (though I'd have to polyfill for the browser).

**Comment by mudge** [comment]
@rbelouin I've published Pacta 0.9 which has the new (now undocumented) `mapError` implemented purely in terms of `onRejected`. In doing so, I think I finally spotted the issue you were having with chaining `onRejected`: it would lose the successful state of a promise:

``` javascript
Promise.resolve(1).onRejected(function () { return 2; }).map(console.log); //=> should be 1
```

This is also fixed in 0.9.

**Comment by rbelouin** [comment]
Nice :)
